pkg/ocicni/ocicni.go

```diff
@@ -174,28 +183,116 @@ func (plugin *cniNetworkPlugin) Name() string {
 	return CNIPluginName
 }
 
-func (plugin *cniNetworkPlugin) SetUpPod(netnsPath string, namespace string, name string, id string) error {
+func (plugin *cniNetworkPlugin) monitorNetDir(netnsPath string, namespace string, name string, id string, timeout time.Duration, cb SetUpCallback, data interface{}) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		logrus.Errorf("could not create new watcher %v", err)
+		cb(data, err)
+		return
+	}
+	defer watcher.Close()
+
+	monitorTimeout := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				logrus.Debugf("CNI monitoring event %v", event)
+				if event.Op&fsnotify.Create != fsnotify.Create {
+					continue
+				}
+
+				if err1 := plugin.setUpPod(netnsPath, namespace, name, id); err1 != nil {
+					logrus.Errorf("could not set up pod with new CNI conf %v", err1)
+					plugin.terminateMonitorNetDir(err1)
+				}
+
+				logrus.Debugf("CNI asynchronous setting succeeded")
+				plugin.terminateMonitorNetDir(nil)
+				return
+
+			case err2 := <-watcher.Errors:
+				logrus.Errorf("CNI monitoring error %v", err2)
+				plugin.terminateMonitorNetDir(err2)
+				return
+
+			case <-monitorTimeout:
+				logrus.Errorf("CNI monitoring timeout")
+				return
+			}
+		}
+	}()
+
+	if err = watcher.Add(plugin.pluginDir); err != nil {
+		logrus.Error(err)
+		cb(data, err)
+		return
+	}
+
+	select {
+	case err = <-plugin.monitorNetDirChan:
+		logrus.Debugf("CNI monitoring finished")
+		cb(data, err)
+
+	case <-time.After(time.Second * timeout):
+		close(monitorTimeout)
+		cb(data, fmt.Errorf("CNI monitoring timeout"))
+	}
+}
+
+func (plugin *cniNetworkPlugin) setUpPod(netnsPath string, namespace string, name string, id string) error {
+	plugin.syncNetworkConfig()
+
 	if err := plugin.checkInitialized(); err != nil {
 		return err
 	}
 
 	_, err := plugin.loNetwork.addToNetwork(name, namespace, id, netnsPath)
 	if err != nil {
-		glog.Errorf("Error while adding to cni lo network: %s", err)
+		logrus.Errorf("Error while adding to cni lo network: %s", err)
 		return err
 	}
 
 	_, err = plugin.getDefaultNetwork().addToNetwork(name, namespace, id, netnsPath)
 	if err != nil {
-		glog.Errorf("Error while adding to cni network: %s", err)
+		logrus.Errorf("Error while adding to cni network: %s", err)
 		return err
 	}
 
 	return err
+
+}
+
+func (plugin *cniNetworkPlugin) SetUpPod(netnsPath string, namespace string, name string, id string, cb SetUpCallback, data interface{}) error {
+	// First let's sync with the latest configuration files
+	plugin.syncNetworkConfig()
```

Comment:
Doesn't this mean that syncNetworkConfig will be called for every pod, regardless of whether the network has already been set up? If I'm following the code correctly, doesn't that mean that the CNI config will be loaded/parsed whenever a pod is created?